### PR TITLE
authz: Add opt in for repository permissions checking for site admins

### PR DESCRIPTION
**Pull request by ryanslade**
By default, site admins can see all code.

This change introduces the authz.enforceForSiteAdmins setting which enforces
permission checking for site admins too.

Closes: https://github.com/sourcegraph/sourcegraph/issues/20622

### CODE REVIEW

**Comment by ryanslade** [comment]
@unknwon This change seems a little too simple 😄 

Is there anything else we need to do? Do you think this could cause issues for our customers?

File: internal/database/repos_perm.go [hunk 1]
```diff
@@ -61,16 +62,18 @@ func AuthzQueryConds(ctx context.Context, db dbutil.DB) (*sqlf.Query, error) {
 
 	authenticatedUserID := int32(0)
 
-	// Authz is bypassed when the request is coming from an internal actor or
-	// there is no authz provider configured and access to all repositories are allowed by default.
+	// Authz is bypassed when the request is coming from an internal actor or there
+	// is no authz provider configured and access to all repositories are allowed by
+	// default. Authz can be bypassed by site admins unless
+	// conf.AuthEnforceForSiteAdmins is set.
```
Comment (unknwon) [inline internal/database/repos_perm.go:68]:
```suggestion
	// conf.AuthEnforceForSiteAdmins is set to "true".
```

File: schema/site.schema.json [hunk 2]
```diff
@@ -528,6 +528,11 @@
       ],
       "group": "Security"
     },
+    "authz.enforceForSiteAdmins": {
+      "description": "When true, site admins will only be able to see private code they have access to via our authz system",
```
Comment (unknwon) [inline schema/site.schema.json:532]:
```suggestion
      "description": "When true, site admins will only be able to see private code they have access to via our authz system.",
```

File: internal/database/repos_perm_test.go [hunk 3]
```diff
@@ -328,7 +355,24 @@ VALUES
 		t.Fatalf("Mismatch (-want +got):\n%s", diff)
 	}
 
+	// When AuthzEnforceForSiteAdmins is set, site admins can only see repos they have access
+	// to based on our authz model
+	conf.Get().AuthzEnforceForSiteAdmins = true
```
Comment (unknwon) [inline internal/database/repos_perm_test.go:360]: TIL this works ...

**Review by unknwon — APPROVED** [review]
LGTM!

**Pull request merged**